File: examples/generator-composition.js
#!/usr/bin/env bun

// Simulated async data source
async function* streamData() {
  const messages = ['hello', 'world', 'from', 'async', 'generators'];
  for (const msg of messages) {
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, 100));
    yield msg
    yield " "
  }
}

// Split strings into individual characters
async function* chars(strings) {
  for await (const word of strings) {
    yield* word;  // yield* works perfectly with async!
  }
}

// Transform characters to uppercase
async function* upper(chars) {
  for await (const c of chars) {
    yield c.toUpperCase();
  }
}

// Add delays between characters for effect
async function* slowType(chars, ms = 50) {
  for await (const c of chars) {
    yield c;
    await new Promise(resolve => setTimeout(resolve, ms));
  }
}

// Main - compose all the generators
async function main() {
  console.log('Streaming characters:\n');
  
  // Natural composition - just wrap generators
  const pipeline = slowType(upper(chars(streamData())));
  
  // Simple consumption
  for await (const c of pipeline) {
    process.stdout.write(c);
  }
  
  console.log('\n\nDone!');
}

// Run it
main().catch(console.error);

File: examples/generator-composition.py
#!/usr/bin/env python3

import asyncio
from typing import AsyncIterator, AsyncGenerator

# Simulated async data source
async def stream_data() -> AsyncGenerator[str, None]:
    messages = ['hello', 'world', 'from', 'async', 'generators']
    for msg in messages:
        # Simulate network delay
        await asyncio.sleep(0.1)
        yield msg
        yield " "

# Split strings into individual characters
async def chars(strings: AsyncIterator[str]) -> AsyncGenerator[str, None]:
    async for s in strings:
        # Can't use yield from with async generators!
        for c in s:
            yield c

# Transform characters to uppercase
async def upper(chars: AsyncIterator[str]) -> AsyncGenerator[str, None]:
    async for c in chars:
        yield c.upper()

# Add delays between characters for effect
async def slow_type(chars: AsyncIterator[str], seconds: float = 0.05) -> AsyncGenerator[str, None]:
    async for c in chars:
        yield c
        await asyncio.sleep(seconds)

# Main - compose all the generators
async def main():
    print('Streaming characters:\n')
    
    # Composition requires explicit generator management
    stream_gen = stream_data()
    chars_gen = chars(stream_gen)
    upper_gen = upper(chars_gen)
    pipeline = slow_type(upper_gen)
    
    try:
        # Consume the pipeline
        async for c in pipeline:
            print(c, end='', flush=True)
    finally:
        # Must explicitly close all generators!
        await pipeline.aclose()
        await upper_gen.aclose()
        await chars_gen.aclose()
        await stream_gen.aclose()
    
    print('\n\nDone!')

# Can't just call main() - need event loop
if __name__ == '__main__':
    asyncio.run(main())
